File: benchmarks/binarytrees/ref.py
# binarytrees reference, Python. Tuples for nodes; None is a leaf.
import sys

sys.setrecursionlimit(100000)


def build(d):
    if d == 0:
        return None
    return (build(d - 1), build(d - 1))


def check(t):
    if t is None:
        return 1
    return 1 + check(t[0]) + check(t[1])


maxdepth = int(sys.argv[1]) if len(sys.argv) > 1 else 14
longlived = build(maxdepth)

d = 4
while d <= maxdepth:
    iters = 1 << (maxdepth - d + 4)
    acc = 0
    for _ in range(iters):
        acc += check(build(d))
    print("%d trees of depth %d check %d" % (iters, d, acc))
    d += 2

print("long-lived tree of depth %d check %d" % (maxdepth, check(longlived)))

File: benchmarks/binarytrees/ref.rs
// binarytrees reference, Rust. Box per node, dropped when the tree leaves scope.
use std::env;

enum Tree {
    Leaf,
    Branch(Box<Tree>, Box<Tree>),
}

fn build(d: i32) -> Tree {
    if d == 0 { Tree::Leaf } else { Tree::Branch(Box::new(build(d - 1)), Box::new(build(d - 1))) }
}

fn check(t: &Tree) -> i64 {
    match t {
        Tree::Leaf => 1,
        Tree::Branch(l, r) => 1 + check(l) + check(r),
    }
}

fn main() {
    let maxdepth: i32 = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(14);
    let longlived = build(maxdepth);
    let mut d = 4;
    while d <= maxdepth {
        let iters: i64 = 1i64 << (maxdepth - d + 4);
        let mut acc: i64 = 0;
        for _ in 0..iters {
            acc += check(&build(d));
        }
        println!("{} trees of depth {} check {}", iters, d, acc);
        d += 2;
    }
    println!("long-lived tree of depth {} check {}", maxdepth, check(&longlived));
}
